crates/ide-completion/src/context.rs

```diff
@@ -569,6 +569,28 @@ impl<'a> CompletionContext<'a> {
         // completing on
         let original_token = original_file.syntax().token_at_offset(offset).left_biased()?;
 
+        // try to skip completions on path with qinvalid colons
+        // this approach works in normal path and inside token tree
+        match original_token.kind() {
+            T![:] => {
+                // return if no prev token before colon
+                let prev_token = original_token.prev_token()?;
+
+                // only has a single colon
+                if prev_token.kind() != T![:] {
+                    return None;
+                }
+
+                if !is_prev_token_valid_path_start_or_segment(&prev_token) {
+                    return None;
+                }
+            }
```

Comment:
I think we can just return None if we find a `T![:]` without any additional checks? The previous token can't be another `T![:]` as the parser should produce a `T![::]` then I think?

Reply:
Parser will produce two continuous `T![:]` if we are inside token streams and that's the case I wanted to support as the video in the description.

Reply:
However that will produce too much noise if we are accidentally typing like `:::`. Every single `:` after will trigger the completion again. So I introduced `is_prev_token_valid_path_start_or_segment` to solve that. I know it's invalid input but felt this behavior pretty annoying during testing without the check.

Reply:
Oh I see, ye then we need something like that here. Though I think this is still too restrictive, so maybe let's instead check that we either have a `::`, or exactly two `:` tokens in a row (but not three)? I think that's a decent trade off unless I am missing something else.

Reply:
updated in b701b14ca5b195fd6e227b913a912a4fcc87efb0